### ISSUE TITLE
Error on PS4 with il2cpp

### ISSUE THREAD

**Issue by d11ahirst** [getsentry/sentry-unity#233]
### Environment

Plugin 0.2.0

How do you use Sentry?
Onpremise

Which SDK and version?
Net 2.0.0

### Steps to Reproduce

Initialised sentry plugin on PS4 Console on boot

### Expected Result

No Errors

What you thought would happen.

No Errors

### Actual Result

  Debug: Logging enabled with ConsoleDiagnosticLogger and min level: Debug
ArgumentOutOfRangeException: The added or subtracted value results in an un-representable DateTime.
Parameter name: value
  at Sentry.Internal.ProcessInfo..ctor (Sentry.SentryOptions options, System.Func`1[TResult] findPreciseStartupTime) [0x00000] in <00000000000000000000000000000000>:0
  at Sentry.SentrySdk.InitHub (Sentry.SentryOptions options) [0x00000] in <00000000000000000000000000000000>:0
  at Sentry.SentrySdk.Init (Sentry.SentryOptions options) [0x00000] in <00000000000000000000000000000000>:0
  at Bootstrap.Start () [0x00000] in <00000000000000000000000000000000>:0

(Filename: currently not available on il2cpp Line: -1)

What actually happened. Maybe a screenshot/recording? Maybe some logs?

**Comment by bruno-garcia** [comment]
@d11ahirst what version of the SDK are you using?
I believe this was resolved on version 0.0.8 already: https://github.com/getsentry/sentry-unity/blob/main/CHANGELOG.md#fixes-8

`ProcessInfo` is not supported on IL2CPP so we turned that feature off for Unity.

**Comment by d11ahirst** [comment]
Sorry for the confusion, we are using Unity 2019.3 with il2cpp on the PS4

**Comment by semuserable** [comment]
Very interesting... My assumption. 

Does it mean that `DateTimeOffset.UtcNow` gets something like `01/01/0001` (ie `DateTimeOffset.MinValue`) on `PS4`? Sentry .NET "substracts" time in `ProcessInfo` during hub creation.

```csharp
var now = DateTimeOffset.UtcNow;
StartupTime = now;
try
{
    BootTime = now.AddTicks(-Stopwatch.GetTimestamp()
                            / (Stopwatch.Frequency
                               / TimeSpan.TicksPerSecond));
}
```

If we try to substract something from `DateTimeOffset.MinValue`, we recieve the mentioned exception.

**Comment by bruno-garcia** [comment]
We rely on `DateTimeOffset.UtcNow` to add a timestamp to the Sentry Event and crumbs and other parts of the SDK so it would be very surprising if that returned `DateTimeOffset.MinValue`.

I think it's more likely this returns it min value:
https://github.com/getsentry/sentry-dotnet/blob/f84c23b9683c8d90880d613701fd9359528061bb/src/Sentry/Internal/ProcessInfo.cs#L93-L94

But that's the part that shouldn't be called on Unity because `DetectStartupTime` is not `Best`:

https://github.com/getsentry/sentry-unity/blob/f1c77e2286cc8f00c20dcd4fb042a26e25aedd23/src/Sentry.Unity/SentryUnityOptions.cs#L85-L86

I'll make sure that failing to instantiate `ProcessInfo` won't bubble out of the SDK so "worst case" we don't have that timestamp but things work fine.

**Comment by bruno-garcia** [comment]
@d11ahirst could you please try initializing the SDK programatically and set `options.DetectStartupTime  = StartupTimeDetectionMode.None` ? 

Without the line number we can only guess and I just want to make sure my patch will work.

**Comment by bruno-garcia** [comment]
Opened this to address the issue: https://github.com/getsentry/sentry-dotnet/pull/1062/

**Comment by d11ahirst** [comment]
@bruno-garcia 

I tried  `options.DetectStartupTime = StartupTimeDetectionMode.None` and it seems to be getting past it.

However sentry is now returning this

```
Sentry: Error
An error occurred when capturing the event UnityEngine.UnityException: GetOperatingSystem can only be called from the main thread.
Constructors and field initializers will be executed from the loading thread when loading a scene.
Don't use this function in the constructor or field initializers, instead move initialization code to the Awake or Start function.
  at (wrapper managed-to-native) UnityEngine.EditorSystemInfo.GetOperatingSystem()
  at UnityEngine.EditorSystemInfo.get_operatingSystem () [0x00001] in <8ee633f6dccc459bbde4324d82f7c465>:0 
  at UnityEngine.SystemInfoShimBase.get_operatingSystem () [0x00000] in <8ee633f6dccc459bbde4324d82f7c465>:0 
  at UnityEngine.SystemInfo.get_operatingSystem () [0x00005] in <8ee633f6dccc459bbde4324d82f7c465>:0 
  at Sentry.Unity.UnityEventProcessor.Process (Sentry.SentryEvent event) [0x00035] in D:\a\sentry-unity\sentry-unity\src\Sentry.Unity\UnityEventProcessor.cs:37 
  at Sentry.SentryClient.DoSendEvent (Sentry.SentryEvent event, Sentry.Scope scope) [0x001cb] in <c15f2d7635c44e498f5ab4870f374e64>:0 
  at Sentry.SentryClient.CaptureEvent (Sentry.SentryEvent event, Sentry.Scope scope) [0x0001f] in <c15f2d7635c44e498f5ab4870f374e64>:0 .

UnityEngine.Debug:LogError(Object)
Sentry.Unity.UnityLogger:Log(SentryLevel, String, Exception, Object[]) (at D:/a/sentry-unity/sentry-unity/src/Sentry.Unity/UnityLogger.cs:42)
Sentry.Extensibility.DiagnosticLoggerExtensions:LogIfEnabled(IDiagnosticLogger, SentryLevel, String, Exception, SentryId, Exception)
Sentry.Extensibility.DiagnosticLoggerExtensions:LogError(IDiagnosticLogger, String, Exception, SentryId)
Sentry.SentryClient:CaptureEvent(SentryEvent, Scope)
Sentry.Internal.Hub:CaptureEvent(SentryEvent, Scope)
Sentry.SentryClientExtensions:CaptureException(ISentryClient, Exception)
Sentry.Integrations.TaskUnobservedTaskExceptionIntegration:Handle(Object, UnobservedTaskExceptionEventArgs)
Sentry.Internal.AppDomainAdapter:OnUnobservedTaskException(Object, UnobservedTaskExceptionEventArgs)
System.Threading.Tasks.TaskExceptionHolder:Finalize()
```

**Comment by bruno-garcia** [comment]
@d11ahirst Work around for that is to call `options.DisableTaskUnobservedTaskExceptionCapture()`

But good point, this also means that a you call `Capture` outside of the main thread it can fail with the same error. We'll address this.

**Comment by bruno-garcia** [comment]
@d11ahirst if you'd like, we can chat on Discord (we have a #unity channel there), Sentry's Discord: https://discord.gg/sentry
I'd also be glad to hope on a call with you and see this through if you're interested.

**Comment by semuserable** [comment]
Should we cache the values from `SystemInfo` during startup? As far as I can tell, UNity doesn't cache it on C# level, most (all?) properties call to `extern`. 

Example:

```csharp
[FreeFunction("systeminfo::GetOperatingSystemFamily")]
[MethodImpl(MethodImplOptions.InternalCall)]
private static extern OperatingSystemFamily GetOperatingSystemFamily();
```

**Comment by bruno-garcia** [comment]
Yes but at the same time if we're eager calling all of these APIs we'll slow down the startup of the game so lets explore other avenues where we can call these APIs asynchronously but running on the main thread. Any event captured before these APIs are called and cached would just not include that data (or if the capture is called from the UI thread we could block while we run the APIs and cache the data for the first time).

**Comment by bruno-garcia** [comment]
Issues described here were resolved on 0.3.0. Please let us know if it works out for you.
